diff --git a/src/inference/batch.jl b/src/inference/batch.jl
index a619c90b..a9145db1 100644
--- a/src/inference/batch.jl
+++ b/src/inference/batch.jl
@@ -23,8 +23,8 @@ struct InferenceResult{P, A, F, M, E}
     error       :: E
 end
 
-Base.iterate(results::InferenceResult)      = iterate((getfield(results, :posteriors), getfield(results, :predictions), getfield(results, :free_energy), getfield(results, :model), getfield(results, :returnval), getfield(results, :error)))
-Base.iterate(results::InferenceResult, any) = iterate((getfield(results, :posteriors), getfield(results, :predictions), getfield(results, :free_energy), getfield(results, :model), getfield(results, :returnval), getfield(results, :error)), any)
+Base.iterate(results::InferenceResult)      = iterate((getfield(results, :posteriors), getfield(results, :predictions), getfield(results, :free_energy), getfield(results, :model), getfield(results, :error)))
+Base.iterate(results::InferenceResult, any) = iterate((getfield(results, :posteriors), getfield(results, :predictions), getfield(results, :free_energy), getfield(results, :model), getfield(results, :error)), any)
 
 """
 Checks if the `InferenceResult` object does not contain an error. 
