diff --git a/src/model_macro.jl b/src/model_macro.jl
index 5a1388f..d138579 100644
--- a/src/model_macro.jl
+++ b/src/model_macro.jl
@@ -174,22 +174,22 @@ Save the expression found in the tilde syntax in the `created_by` field of the e
 function save_expression_in_tilde(e::Expr)
     if @capture(e, (local lhs_ ~ rhs_ where {options__}) | (local lhs_ ~ rhs_))
         options = options === nothing ? [] : options
-        return :(local $lhs ~ $rhs where {$(options...), created_by = () -> $(Expr(:quote, prettify(e)))})
+        return :(local $lhs ~ $rhs where {$(options...), created_by = $(QuoteNode(prettify(e)))})
     elseif @capture(e, (local lhs_ .~ rhs_ where {options__}) | (local lhs_ .~ rhs_))
         options = options === nothing ? [] : options
-        return :(local $lhs .~ $rhs where {$(options...), created_by = () -> $(Expr(:quote, prettify(e)))})
+        return :(local $lhs .~ $rhs where {$(options...), created_by = $(QuoteNode(prettify(e)))})
     elseif @capture(e, (local lhs_ := rhs_ where {options__}) | (local lhs_ := rhs_))
         options = options === nothing ? [] : options
-        return :(local $lhs := $rhs where {$(options...), created_by = () -> $(Expr(:quote, prettify(e)))})
+        return :(local $lhs := $rhs where {$(options...), created_by = $(QuoteNode(prettify(e)))})
     elseif @capture(e, (lhs_ ~ rhs_ where {options__}) | (lhs_ ~ rhs_))
         options = options === nothing ? [] : options
-        return :($lhs ~ $rhs where {$(options...), created_by = () -> $(Expr(:quote, prettify(e)))})
+        return :($lhs ~ $rhs where {$(options...), created_by = $(QuoteNode(prettify(e)))})
     elseif @capture(e, (lhs_ .~ rhs_ where {options__}) | (lhs_ .~ rhs_))
         options = options === nothing ? [] : options
-        return :($lhs .~ $rhs where {$(options...), created_by = () -> $(Expr(:quote, prettify(e)))})
+        return :($lhs .~ $rhs where {$(options...), created_by = $(QuoteNode(prettify(e)))})
     elseif @capture(e, (lhs_ := rhs_ where {options__}) | (lhs_ := rhs_))
         options = options === nothing ? [] : options
-        return :($lhs := $rhs where {$(options...), created_by = () -> $(Expr(:quote, prettify(e)))})
+        return :($lhs := $rhs where {$(options...), created_by = $(QuoteNode(prettify(e)))})
     else
         return e
     end
diff --git a/test/model_macro_tests.jl b/test/model_macro_tests.jl
index 30d53bd..0e9715d 100644
--- a/test/model_macro_tests.jl
+++ b/test/model_macro_tests.jl
@@ -224,7 +224,7 @@ end
 
     # Test 1: save expression in tilde
     input = :(x ~ Normal(0, 1))
-    output = :(x ~ Normal(0, 1) where {created_by = () -> :(x ~ Normal(0, 1))})
+    output = :(x ~ Normal(0, 1) where {created_by = $(QuoteNode(:(x ~ Normal(0, 1))))})
     @test_expression_generating save_expression_in_tilde(input) output
 
     # Test 2: save expression in tilde with multiple expressions
@@ -233,14 +233,14 @@ end
         y ~ Normal(0, 1)
     end
     output = quote
-        x ~ Normal(0, 1) where {created_by = () -> :(x ~ Normal(0, 1))}
-        y ~ Normal(0, 1) where {created_by = () -> :(y ~ Normal(0, 1))}
+        x ~ Normal(0, 1) where {created_by = $(QuoteNode(:(x ~ Normal(0, 1))))}
+        y ~ Normal(0, 1) where {created_by = $(QuoteNode(:(y ~ Normal(0, 1))))}
     end
     @test_expression_generating apply_pipeline(input, save_expression_in_tilde) output
 
     # Test 3: save expression in tilde with broadcasted operation
     input = :(x .~ Normal(0, 1))
-    output = :(x .~ Normal(0, 1) where {created_by = () -> :(x .~ Normal(0, 1))})
+    output = :(x .~ Normal(0, 1) where {created_by = $(QuoteNode(:(x .~ Normal(0, 1))))})
     @test_expression_generating save_expression_in_tilde(input) output
 
     # Test 4: save expression in tilde with multiple broadcast expressions
@@ -250,15 +250,15 @@ end
     end
 
     output = quote
-        x .~ Normal(0, 1) where {created_by = () -> :(x .~ Normal(0, 1))}
-        y .~ Normal(0, 1) where {created_by = () -> :(y .~ Normal(0, 1))}
+        x .~ Normal(0, 1) where {created_by = $(QuoteNode(:(x .~ Normal(0, 1))))}
+        y .~ Normal(0, 1) where {created_by = $(QuoteNode(:(y .~ Normal(0, 1))))}
     end
 
     @test_expression_generating apply_pipeline(input, save_expression_in_tilde) output
 
     # Test 5: save expression in tilde with deterministic operation
     input = :(x := Normal(0, 1))
-    output = :(x := Normal(0, 1) where {created_by = () -> :(x := Normal(0, 1))})
+    output = :(x := Normal(0, 1) where {created_by = $(QuoteNode(:(x := Normal(0, 1))))})
     @test_expression_generating save_expression_in_tilde(input) output
 
     # Test 6: save expression in tilde with multiple deterministic expressions
@@ -268,8 +268,8 @@ end
     end
 
     output = quote
-        x := Normal(0, 1) where {created_by = () -> :(x := Normal(0, 1))}
-        y := Normal(0, 1) where {created_by = () -> :(y := Normal(0, 1))}
+        x := Normal(0, 1) where {created_by = $(QuoteNode(:(x := Normal(0, 1))))}
+        y := Normal(0, 1) where {created_by = $(QuoteNode(:(y := Normal(0, 1))))}
     end
 
     @test_expression_generating apply_pipeline(input, save_expression_in_tilde) output
@@ -280,29 +280,29 @@ end
         y ~ Normal(0, 1) where {q = MeanField()}
     end
     output = quote
-        x ~ Normal(0, 1) where {q = MeanField(), created_by = () -> :(x ~ Normal(0, 1) where {q = MeanField()})}
-        y ~ Normal(0, 1) where {q = MeanField(), created_by = () -> :(y ~ Normal(0, 1) where {q = MeanField()})}
+        x ~ Normal(0, 1) where {q = MeanField(), created_by = $(QuoteNode(:(x ~ Normal(0, 1) where {q = MeanField()})))}
+        y ~ Normal(0, 1) where {q = MeanField(), created_by = $(QuoteNode(:(y ~ Normal(0, 1) where {q = MeanField()})))}
     end
     @test_expression_generating apply_pipeline(input, save_expression_in_tilde) output
 
     # Test 8: with different variable names
     input = :(y ~ Normal(0, 1))
-    output = :(y ~ Normal(0, 1) where {created_by = () -> :(y ~ Normal(0, 1))})
+    output = :(y ~ Normal(0, 1) where {created_by = $(QuoteNode(:(y ~ Normal(0, 1))))})
     @test_expression_generating save_expression_in_tilde(input) output
 
     # Test 9: with different parameter options
     input = :(x ~ Normal(0, 1) where {mu = 2.0, sigma = 0.5})
-    output = :(x ~ Normal(0, 1) where {mu = 2.0, sigma = 0.5, created_by = () -> :(x ~ Normal(0, 1) where {mu = 2.0, sigma = 0.5})})
+    output = :(x ~ Normal(0, 1) where {mu = 2.0, sigma = 0.5, created_by = $(QuoteNode(:(x ~ Normal(0, 1) where {mu = 2.0, sigma = 0.5})))})
     @test_expression_generating save_expression_in_tilde(input) output
 
     # Test 10: with different parameter options
     input = :(y ~ Normal(0, 1) where {mu = 1.0})
-    output = :(y ~ Normal(0, 1) where {mu = 1.0, created_by = () -> :(y ~ Normal(0, 1) where {mu = 1.0})})
+    output = :(y ~ Normal(0, 1) where {mu = 1.0, created_by = $(QuoteNode(:(y ~ Normal(0, 1) where {mu = 1.0})))})
     @test_expression_generating save_expression_in_tilde(input) output
 
     # Test 11: with no parameter options
     input = :(x ~ Normal(0, 1) where {})
-    output = :(x ~ Normal(0, 1) where {created_by = () -> :(x ~ Normal(0, 1) where {})})
+    output = :(x ~ Normal(0, 1) where {created_by = $(QuoteNode(:(x ~ Normal(0, 1) where {})))})
 
     # Test 12: check unmatching pattern
     input = quote
@@ -322,7 +322,7 @@ end
     end
     output = quote
         for i in 1:10
-            x[i] ~ Normal(0, 1) where {created_by = () -> :(x[i] ~ Normal(0, 1))}
+            x[i] ~ Normal(0, 1) where {created_by = $(QuoteNode(:(x[i] ~ Normal(0, 1))))}
         end
     end
     @test_expression_generating save_expression_in_tilde(input) input
@@ -334,8 +334,8 @@ end
     end
 
     output = quote
-        local x ~ (Normal(0, 1)) where {created_by = () -> :(local x ~ Normal(0, 1))}
-        local y ~ (Normal(0, 1)) where {created_by = () -> :(local y ~ Normal(0, 1))}
+        local x ~ (Normal(0, 1)) where {created_by = $(QuoteNode(:(local x ~ Normal(0, 1))))}
+        local y ~ (Normal(0, 1)) where {created_by = $(QuoteNode(:(local y ~ Normal(0, 1))))}
     end
 
     @test_expression_generating save_expression_in_tilde(input) input
@@ -345,7 +345,7 @@ end
         x := a + b
     end
     output = quote
-        x := (a + b) where {created_by = () -> :(x := a + b)}
+        x := (a + b) where {created_by = $(QuoteNode(:(x := a + b)))}
     end
     @test_expression_generating save_expression_in_tilde(input) output
 
@@ -354,7 +354,7 @@ end
         local x := a + b
     end
     output = quote
-        local x := (a + b) where {created_by = () -> :(local x := a + b)}
+        local x := (a + b) where {created_by = $(QuoteNode(:(local x := a + b)))}
     end
     @test_expression_generating save_expression_in_tilde(input) output
 
@@ -363,7 +363,7 @@ end
         local x := (a + b) where {q = q(x)q(a)q(b)}
     end
     output = quote
-        local x := (a + b) where {q = q(x)q(a)q(b), created_by = () -> :(local x := (a + b) where {q = q(x)q(a)q(b)})}
+        local x := (a + b) where {q = q(x)q(a)q(b), created_by = $(QuoteNode(:(local x := (a + b) where {q = q(x)q(a)q(b)})))}
     end
     @test_expression_generating save_expression_in_tilde(input) output
 
@@ -372,7 +372,7 @@ end
         local x .~ Normal(μ, σ)
     end
     output = quote
-        local x .~ Normal(μ, σ) where {created_by = () -> :(local x .~ Normal(μ, σ))}
+        local x .~ Normal(μ, σ) where {created_by = $(QuoteNode(:(local x .~ Normal(μ, σ))))}
     end
     @test_expression_generating save_expression_in_tilde(input) output
 
@@ -381,7 +381,7 @@ end
         local x .~ Normal(μ, σ) where {q = q(x)q(μ)q(σ)}
     end
     output = quote
-        local x .~ Normal(μ, σ) where {q = q(x)q(μ)q(σ), created_by = () -> :(local x .~ Normal(μ, σ) where {q = q(x)q(μ)q(σ)})}
+        local x .~ Normal(μ, σ) where {q = q(x)q(μ)q(σ), created_by = $(QuoteNode(:(local x .~ Normal(μ, σ) where {q = q(x)q(μ)q(σ)})))}
     end
     @test_expression_generating save_expression_in_tilde(input) output
 end
